// flagrate/lib/es6/flagrate/progress.js
/*
   Copyright 2016 Webnium

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
"use strict";
import { extendObject } from "./util";
import { Element } from "./element";
/*?
    flagrate.createProgress(option)
    new flagrate.Progress(option)
    - option (Object) - options.
**/
function FProgress(opt = {}) {
    const attr = opt.attribute || {};
    attr["id"] = opt.id || null;
    attr["class"] = opt.className || null;
    //create
    const progress = new Element("div", attr);
    extendObject(progress, this);
    progress._value = opt.value || 0;
    progress._max = opt.max || 100;
    progress.addClassName("flagrate flagrate-progress");
    progress._bar = new Element().insertTo(progress);
    progress._updateProgress();
    return progress;
}
export const Progress = FProgress;
export function createProgress(option) {
    return new Progress(option);
}
Progress.prototype = {
    getValue() {
        return this._value;
    },
    setValue(number) {
        const progress = this;
        if (typeof number !== "number") {
            return progress;
        }
        progress._value = Math.max(0, Math.min(progress._max, number));
        progress.fire("change", { targetProgress: progress });
        progress._updateProgress();
        return progress;
    },
    getMax() {
        return this._max;
    },
    setMax(number) {
        const progress = this;
        if (typeof number !== "number") {
            return progress;
        }
        progress._max = number;
        progress.setValue(progress._value);
        progress._updateProgress();
        return progress;
    },
    _updateProgress() {
        const progress = this;
        const percentage = Math.max(0, Math.min(100, progress._value / progress._max * 100));
        progress._bar.setStyle({ width: `${percentage}%` });
        return;
    }
};
