// flagrate/lib/es6/flagrate/switch.js
/*
   Copyright 2016 Webnium

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
"use strict";
import { extendObject } from "./util";
import { Button } from "./button";
/*?
    flagrate.createSwitch(option)
    new flagrate.Switch(option)
    - option (Object) - options.
**/
function FSwitch(opt = {}) {
    //create
    const sw = new Button({
        id: opt.id,
        className: opt.className,
        attribute: opt.attribute,
        style: opt.style,
        isFocused: opt.isFocused,
        isDisabled: opt.isDisabled
    });
    extendObject(sw, this);
    sw.addEventListener("select", sw.toggleSwitch.bind(sw));
    sw.addClassName("flagrate-switch");
    if (sw.dataset) {
        sw.dataset["flagrateSwitchStatus"] = opt.isOn ? "on" : "off";
    }
    else {
        sw.writeAttribute("data-flagrate-switch-status", opt.isOn ? "on" : "off");
    }
    return sw;
}
export const Switch = FSwitch;
export function createSwitch(option) {
    return new Switch(option);
}
Switch.prototype = {
    isOn() {
        if (this.dataset) {
            return this.dataset.flagrateSwitchStatus === "on";
        }
        else {
            return this.readAttribute("data-flagrate-switch-status") === "on";
        }
    },
    switchOn() {
        if (this.dataset) {
            this.dataset.flagrateSwitchStatus = "on";
        }
        else {
            this.writeAttribute("data-flagrate-switch-status", "on");
        }
        return this.fire("on", { targetSwitch: this })
            .fire("change", { targetSwitch: this });
    },
    switchOff() {
        if (this.dataset) {
            this.dataset.flagrateSwitchStatus = "off";
        }
        else {
            this.writeAttribute("data-flagrate-switch-status", "off");
        }
        return this.fire("off", { targetSwitch: this })
            .fire("change", { targetSwitch: this });
    },
    toggleSwitch() {
        return this.isOn() ? this.switchOff() : this.switchOn();
    }
};
