// flagrate/lib/es6/flagrate/text-input.js
/*
   Copyright 2016 Webnium

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
"use strict";
import { extendObject } from "./util";
import { Element } from "./element";
/*?
    flagrate.createTextInput(option)
    new flagrate.TextInput(option)
    - option (Object) - options.

    TextInput.

    #### option

    * `id`                       (String): `id` attribute of `input` element.
    * `className`                (String):
    * `attribute`                (Object):
    * `style`                    (Object): (using flagrate.Element.setStyle)
    * `value`                    (String):
    * `placeholder`              (String):
    * `icon`                     (String):
    * `regexp`                   (RegExp):
    * `isDisabled`               (Boolean; default `false`):
**/
function FTextInput(option = {}) {
    this.regexp = option.regexp || null;
    const attr = option.attribute || {};
    if (option.id) {
        attr["id"] = option.id;
    }
    if (option.value) {
        attr["value"] = option.value;
    }
    if (option.placeholder) {
        attr["placeholder"] = option.placeholder;
    }
    //create
    const input = new Element("input", attr);
    extendObject(input, this);
    input.addClassName("flagrate flagrate-textinput");
    if (option.className) {
        input.addClassName(option.className);
    }
    if (option.style) {
        input.setStyle(option.style);
    }
    if (option.icon) {
        input.setIcon(option.icon);
    }
    if (option.isDisabled) {
        input.disable();
    }
    return input;
}
export const TextInput = FTextInput;
export function createTextInput(option) {
    return new TextInput(option);
}
TextInput.prototype = {
    disable() {
        this.addClassName("flagrate-disabled");
        this.writeAttribute("disabled", true);
        return this;
    },
    enable() {
        this.removeClassName("flagrate-disabled");
        this.writeAttribute("disabled", false);
        return this;
    },
    isEnabled() {
        return !this.hasClassName("flagrate-disabled");
    },
    setValue(value) {
        this.value = value;
        return this;
    },
    getValue() {
        return this.value;
    },
    setIcon(identifier) {
        this._iconIdentifier = identifier;
        if (identifier) {
            return this.addClassName("flagrate-icon").setStyle({
                backgroundImage: `url(${identifier})`
            });
        }
        else {
            return this.removeClassName("flagrate-icon").setStyle({
                backgroundImage: "none"
            });
        }
    },
    getIcon() {
        return this._iconIdentifier || "";
    },
    isValid: function () {
        return this.regexp.test(this.getValue());
    }
};
